File: Sliding Window/1456. Maximum Number of Vowels in a Substring of Given Length/1456. Maximum Number of Vowels in a Substring of Given Length.py
class Solution:
    def maxVowels(self, s: str, k: int) -> int:
        vowels = "aeiou"
        num_vowels = 0

        for char in s[:k]:
            if char in vowels:
                num_vowels += 1

        max_vowels = num_vowels
        
        for i in range(k, len(s)):
            if s[i - k] in vowels:
                num_vowels -= 1
            if s[i] in vowels:
                num_vowels += 1
            
            max_vowels = max(max_vowels, num_vowels)
        
        return max_vowels

# Time: O(n)
# Space: O(1)

File: Sliding Window/1456. Maximum Number of Vowels in a Substring of Given Length/1456. Maximum Number of Vowels in a Substring of Given Length.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var maxVowels = function (s, k) {
  const vowels = "aeiou";
  let num_vowels = 0;

  for (let i = 0; i < k; i++) {
    if (vowels.includes(char)) num_vowels++;
  }

  let max_vowels = num_vowels;

  for (let i = k; i < s.length; i++) {
    if (vowels.includes(s[i - k])) num_vowels--;
    if (vowels.includes(s[i])) num_vowels++;

    max_vowels = Math.max(max_vowels, num_vowels);
  }

  return max_vowels;
};

// Time: O(n)
// Space: O(1)
